Fix trouble to compile with MSVC

annnei opened the pull request:
Remove std::forward here since MSVC does not compile with it. It issues a cryptic error message:

...\duckdb-src\third_party\libpg_query\postgres_parser.cpp(43): error C2243: 'type cast': conversion from 'std::vector<duckdb_libpgquery::PGKeyword,std::allocator<duckdb_libpgquery::PGKeyword>> *' to '_Ty &&' exists, but is inaccessible
          with
          [
              _Ty=duckdb::vector<duckdb_libpgquery::PGKeyword,true>
          ]

But the real problem seems to be that std::forward should not really be used here.

See: https://github.com/duckdb/duckdb/discussions/12487

annnei committed:
Fix trouble to compile with MSVC

Remove std::forward here since MSVC does not compile with it. It issues a
cryptic error message:

...\duckdb-src\third_party\libpg_query\postgres_parser.cpp(43): error C2243: 'type cast': conversion from 'std::vector<duckdb_libpgquery::PGKeyword,std::allocator<duckdb_libpgquery::PGKeyword>> *' to '_Ty &&' exists, but is inaccessible
          with
          [
              _Ty=duckdb::vector<duckdb_libpgquery::PGKeyword,true>
          ]

But the real problem seems to be that std::forward should not really be used here.

carlopi commented:
Thanks for raising / proposing a fix.

I am not sure we could do properly without the forward, given there is a cast in the middle. I need to review when / why it was put there (I think it was past me).

Tishj commented:
I think it was me, I'm also not sure this can just be removed
There is a conversion from `std::vector` to `duckdb::vector` taking place here

annnei commented:
OK, I did not spot the duckdb::vector. But - if a cast is needed, why not use a cast rather than std::forward? What would a forward do here? I think my code gets an implicit cast. 
But my project does not use this code, so I rely on CI testing for this.

Tishj commented:
Maybe something to look into: <https://en.cppreference.com/w/cpp/iterator/move_iterator>

annnei committed:
Fix trouble to compile with MSVC

Remove std::forward here since MSVC does not compile with it. It issues a
cryptic error message:

...\duckdb-src\third_party\libpg_query\postgres_parser.cpp(43): error C2243: 'type cast': conversion from 'std::vector<duckdb_libpgquery::PGKeyword,std::allocator<duckdb_libpgquery::PGKeyword>> *' to '_Ty &&' exists, but is inaccessible
          with
          [
              _Ty=duckdb::vector<duckdb_libpgquery::PGKeyword,true>
          ]

But the real problem seems to be that std::forward should not really be used here.

Switch over to use std::move instead as suggested in https://github.com/duckdb/duckdb/discussions/12487
This does compile with MSVC, but issues a warning: C26479: Don't use std::move to return a local variable

Mytherin commented:
Thanks!

carlopi commented:
Just for completeness, reasoning to have a `std::forward` there was to avoid an (implicit) copy.

Basically we pay a copy to go from std::vector to duckdb::vector, while they are basically the same class.

I am not 100% positive the forward removed the copy, but that was the idea behind it.

We might be better removing `std::move` (that I suggested, but it's not actually needed and flags compilation warning turned errors in the amalgamation). I will send a fixup on that.

carlopi commented:
Also, PR was very very welcome, sorry I was late with reviewing.

Portability comes first, in general and also here, thanks for sending the fix, and keep raising issues and solutions.
